backend-lib/src/main/rust/lib.rs

```diff
@@ -2720,19 +2646,76 @@ fn parse_network(value: u32) -> anyhow::Result<Network> {
     }
 }
 
-fn parse_secret(env: &JNIEnv, secret_bytes: &JByteArray) -> anyhow::Result<SecretVec<u8>> {
+fn path_from_jni(env: &mut JNIEnv, path: JString) -> anyhow::Result<PathBuf> {
+    Ok(PathBuf::from(env.get_string(&path)?.to_str()?))
+}
+
+fn secret_from_jni(env: &JNIEnv, secret_bytes: JByteArray) -> anyhow::Result<SecretVec<u8>> {
     Ok(SecretVec::new(utils::java_bytes_to_rust(
         env,
         &secret_bytes,
     )?))
 }
 
-fn parse_txid(env: &JNIEnv, txid_bytes: &JByteArray) -> anyhow::Result<TxId> {
+fn zip32_account_index_from_jlong(account_index: jlong) -> anyhow::Result<zip32::AccountId> {
+    u32::try_from(account_index)
+        .map_err(|_| ())
+        .and_then(|id| zip32::AccountId::try_from(id).map_err(|_| ()))
+        .map_err(|_| anyhow!("Invalid account ID"))
+}
+
+fn account_id_from_jni(env: &JNIEnv, account_uuid: JByteArray) -> anyhow::Result<AccountUuid> {
+    Ok(AccountUuid::from_uuid(Uuid::from_slice(
+        &utils::java_bytes_to_rust(env, &account_uuid)?,
+    )?))
+}
+
+fn parse_txid(env: &JNIEnv, txid_bytes: JByteArray) -> anyhow::Result<TxId> {
     let txid_bytes = utils::java_bytes_to_rust(env, &txid_bytes)?;
     Ok(TxId::read(&txid_bytes[..])?)
 }
 
-fn parse_pczt(env: &JNIEnv, pczt: &JByteArray) -> anyhow::Result<Pczt> {
+fn parse_treestate(env: &JNIEnv, treestate: JByteArray) -> anyhow::Result<TreeState> {
+    TreeState::decode(utils::java_bytes_to_rust(env, &treestate)?.as_slice())
+        .map_err(|e| anyhow!("Invalid TreeState: {}", e))
+}
+
+fn parse_pczt(env: &JNIEnv, pczt: JByteArray) -> anyhow::Result<Pczt> {
     Pczt::parse(&utils::java_bytes_to_rust(env, &pczt)?)
         .map_err(|e| anyhow!("Invalid PCZT: {:?}", e))
 }
+
+fn parse_ufvk(
+    env: &mut JNIEnv,
+    ufvk_string: JString,
+    network: &Network,
+) -> anyhow::Result<UnifiedFullViewingKey> {
+    let ufvk_string = utils::java_string_to_rust(env, &ufvk_string)?;
+    UnifiedFullViewingKey::decode(network, &ufvk_string)
+        .map_err(|e| anyhow!("Value \"{ufvk_string}\" did not decode as a valid UFVK: {e}"))
+}
+
+struct UnifiedAddressParser(UnifiedAddress);
+
+impl zcash_address::TryFromRawAddress for UnifiedAddressParser {
+    type Error = anyhow::Error;
+
+    fn try_from_raw_unified(
+        data: zcash_address::unified::Address,
+    ) -> Result<Self, zcash_address::ConversionError<Self::Error>> {
+        data.try_into()
+            .map(UnifiedAddressParser)
+            .map_err(|e| anyhow!("Invalid Unified Address: {}", e).into())
+    }
+}
```

Comment:
Why not just use https://github.com/zcash/librustzcash/blob/main/zcash_keys/src/address.rs#L28? 

It's really weird to me to have to go through `TryFromRawAddress` for these straightforward operations.

Reply:
(I mean, why not use it directly?)

Reply:
I _am_ using that; it's the `data.try_into()` call. The rest of the `TryFromRawAddress` machinery here is ensuring that if someone passes a string that isn't a UA (and is instead some other kind of address, or not an address at all), we get correct errors.

Reply:
Ah. So, looking at the call sites of `parse_ua`, what is the reason to restrict these to unified addresses? If we parsed an arbitrary `ZcashAddress` at these call sites, we could still extract the transparent and/or Sapling receivers if present.

Reply:
That might be worth considering, but that's outside the scope of this PR.